plutus-list/src/PlutusTx/List/Natural.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE Trustworthy #-}
+
+{- |
+ Module: PlutusTx.List.Natural
+ Copyright: (C) MLabs 2021
+ License: Apache 2.0
+ Maintainer: Koz Ross <koz@mlabs.city>
+ Portability: GHC only
+ Stability: Experimental
+ List functions that works with the 'Natural' type.
+-}
+module PlutusTx.List.Natural (
+  length,
+  replicate,
+  take,
+  drop,
+  splitAt,
+) where
+
+import Data.Monoid (Sum (Sum, getSum))
+import PlutusTx.Natural (Natural, nat)
+import PlutusTx.Prelude hiding (length, take)
+import Prelude qualified ()
+
+{-# INLINEABLE length #-}
+
+{- | Returns the size/length of a finite structure as a 'Natural'.
+
+@since 1.0
+-}
+length :: Foldable f => f a -> Natural
+length = getSum . foldMap (Sum . const [nat| 1 |])
+
+{-# INLINEABLE replicate #-}
+
+{- | @'replicate' n x@ is a list of length @n@ with @x@ the value of every
+element.
+
+@since 1.0
+-}
+replicate :: Natural -> a -> [a]
+replicate [nat| 0 |] _ = []
+replicate n x = x : replicate (pred n) x
+
+-- [daylily]: these two are not morally "total" enough (uses 'pred'). How to
```

Comment:
This actually falls into the 'use partial function to ensure totality of another function', so it should be OK. You can also use the monus operator `^-` instead if you prefer.

Reply:
I wasn't aware of the monus operator, will change this.